pkg/tekton/metapipeline/metaclient/metaclient.go

```diff
@@ -0,0 +1,324 @@
+		if err != nil {
+			return errors.Wrapf(err, "failed to apply Tekton CRDs")
+		}
+		if o.Verbose {
+			log.Logger().Infof("applied tekton CRDs for %s", tektonCRDs.PipelineRun().Name)
+		}
+	}
+	return nil
+}
+
+func (o *MetaClient) getPodTemplates(kubeClient kubeclient.Interface, ns string, containerName string) (map[string]*corev1.Pod, error) {
+	podTemplates, err := kube.LoadPodTemplates(kubeClient, ns)
+	if err != nil {
+		return nil, err
+	}
+	return podTemplates, nil
+}
+
+// GetFactory returns the factory
+func (o *MetaClient) GetFactory() jxfactory.Factory {
+	if o.factory == nil {
+		o.factory = jxfactory.NewFactory()
+	}
+	return o.factory
+}
+
+func (o *MetaClient) getClientsAndNamespace() (tektonclient.Interface, jxclient.Interface, kubeclient.Interface, string, error) {
+	f := o.GetFactory()
+
+	tektonClient, _, err := f.CreateTektonClient()
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to create Tekton client")
+	}
+
+	jxClient, _, err := f.CreateJXClient()
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to create JX client")
+	}
+
+	kubeClient, ns, err := f.CreateKubeClient()
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to create Kube client")
+	}
+	ns, _, err = kube.GetDevNamespace(kubeClient, ns)
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to find the dev namespace")
+	}
+	return tektonClient, jxClient, kubeClient, ns, nil
+}
+
+// determineBranchIdentifier finds a identifier for the branch which is build by the specified pull ref.
+// The name is either an actual git branch name (eg master) or a synthetic names  like PR-<number> for single pull requests
+// or 'batch' for Prow batch builds.
+// The method makes the decision purely based on the Prow pull ref. At this stage we don't have the full ProwJov spec
+// available.
+func (o *MetaClient) determineBranchIdentifier(pullRef prow.PullRefs) string {
+	prCount := len(pullRef.ToMerge)
+	var branch string
+	switch prCount {
+	case 0:
+		{
+			// no pull requests to merge, taking base branch name as identifier
+			branch = pullRef.BaseBranch
+		}
+	case 1:
+		{
+			// single pull request, create synthetic PR identifier
+			for k := range pullRef.ToMerge {
+				branch = fmt.Sprintf("PR-%s", k)
+				break
+			}
+		}
+	default:
+		{
+			branch = "batch"
+		}
+	}
+	log.Logger().Debugf("branch identifier for pull ref '%s' : '%s'", pullRef.String(), branch)
+	return branch
+}
+
+func (o *MetaClient) determinePipelineKind(pullRef prow.PullRefs) string {
+	var kind string
+
+	prCount := len(pullRef.ToMerge)
+	if prCount > 0 {
+		kind = jenkinsfile.PipelineKindPullRequest
+	} else {
+		kind = jenkinsfile.PipelineKindRelease
+	}
+	log.Logger().Debugf("pipeline kind for pull ref '%s' : '%s'", pullRef.String(), kind)
+	return kind
+}
+
+func (o *MetaClient) gitCloneVersionStream(jxClient jxclient.Interface, ns string) (string, error) {
```

Comment:
Do we really need the pipeline to be dependent to the version stream?

Reply:
yes - we need this for replacing docker image versions inside the meta pipeline